src/primaite/agents/aegis/gllm.py

```diff
@@ -39,105 +37,92 @@ def __init__(
 
         return loss
 
-    def _pad_smallest_texts(self, openai_tokens, gllm_tokens):
-        if openai_tokens.shape[-1] > gllm_tokens.shape[-1]:
-            n_pad = openai_tokens.shape[-1] - gllm_tokens.shape[-1]
-            gllm_tokens = F.pad(gllm_tokens, (0, n_pad), "constant", self.llm.tokenizer.pad_token_id)
-        elif gllm_tokens.shape[-1] > openai_tokens.shape[-1]:
-            n_pad = gllm_tokens.shape[-1] - openai_tokens.shape[-1]
-            openai_tokens = F.pad(openai_tokens, (0, n_pad), "constant", self.llm.tokenizer.pad_token_id)
 
-        return openai_tokens, gllm_tokens
+def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs: int = 10):
+    model.train(mode=True)
 
+    for epoch in range(n_epochs):
+        gllm_responses = []
 
-def train_loop(model, x, edge_index, questions: List[str], network_desc: str):
+        for batch in dataloader:
+            model.optimizer.zero_grad()
+            questions = batch["questions"]
+            gllm_prompts = model.build_prompts(questions=questions, network_desc=network_desc, model="gllm")
 
-    gllm_responses = []
-    openai_responses = []
-    model.train(mode=True)
-    # Get all openai responses now
-    for question in questions:
-        openai_prompt = model._build_prompt(question=question, network_desc=network_desc, model="openai")
-        openai_responses.append(model.openai.generate(prompt=openai_prompt))
+            gllm_texts, gllm_hidden_states = model(batch["graphs"], gllm_prompts)
+            gllm_responses.extend(gllm_texts)
 
-    for episode in range(50):
-        # Reset
-        gllm_responses = []
-        model.optimizer.zero_grad()
-        for idx, question in enumerate(questions):
-            gllm_prompt = model._build_prompt(question=question, network_desc=network_desc, model="gllm")
-            gllm_response = model(x, edge_index, gllm_prompt)
-            gllm_responses.append(gllm_response)
-            
-        loss = model.get_cosine_embeddings_loss(
-            openai_responses, gllm_responses
-        )
+            loss = model.get_cosine_embeddings_loss(batch["gt_hidden_states"], gllm_hidden_states)
+            # Do the backwards pass
+            loss.backward()
+
+            # Clear up any gpu memory that may be holding onto tensors unnecessarily
+            torch.cuda.empty_cache()
+
+        # Print any None gradients
+        # for name, param in model.named_parameters():
+        #     print(f"Gradient for {name}: {param.grad}")
 
-        # Do the backwards pass
-        loss.backward()
-        
-        # Print gradients
-        for name, param in model.named_parameters():
-            print(f"Gradient of {name}: {param.grad}")
-        
         model.optimizer.step()
-        print(f"Episode {episode} Loss:", loss.item())
 
-    return gllm_responses, openai_responses
\ No newline at end of file
+        print(f"Epoch {epoch} Loss:", loss.item(), gllm_responses[0].replace("\n", "//n"))
```

Comment:
This only prints the loss of the last batch in the epoch - just to be aware.

Reply:
Yes I am aware, I'm also printing only 1 response - it's just to give me a general idea of what it's saying :) Thanks